src/Microsoft.ML.TimeSeries/SrCnnBatchAnomalyDetection.cs

```diff
@@ -0,0 +1,778 @@
+        {
+            currentBatch.AddValue(currentInput);
+        }
+
+        protected override void ProcessBatch(Batch currentBatch)
+        {
+            currentBatch.Process();
+            currentBatch.Reset();
+        }
+
+        public sealed class Batch
+        {
+            private List<double> _previousBatch;
+            private List<double> _batch;
+            private readonly int _batchSize;
+            private readonly int _outputLength;
+            private SrCnnEntireModeler _modler;
+            private double[][] _results;
+
+            public Batch(int batchSize, int outputLength, SrCnnEntireModeler modeler)
+            {
+                _batchSize = batchSize;
+                _outputLength = outputLength;
+                _previousBatch = new List<double>(batchSize);
+                _batch = new List<double>(batchSize);
+                _modler = modeler;
+            }
+
+            public void AddValue(double value)
+            {
+                _batch.Add(value);
+            }
+
+            public int Count => _batch.Count;
+
+            public void Process()
+            {
+                if (_batch.Count < MinBatchSize)
+                {
+                    if (_previousBatch.Count + _batch.Count < MinBatchSize)
+                        return;
+                    var bLen = _previousBatch.Count - _batch.Count;
+                    _previousBatch = _previousBatch.GetRange(_batch.Count, bLen);
+                    _previousBatch.AddRange(_batch);
+                    _results = _modler.Train(_previousBatch.ToArray()).Skip(bLen).ToArray();
+                }
+                else
+                {
+                    _results = _modler.Train(_batch.ToArray());
+                }
+            }
+
+            public void Reset()
+            {
+                var tempBatch = _previousBatch;
+                _previousBatch = _batch;
+                _batch = tempBatch;
+                _batch.Clear();
+            }
+
+            public ValueGetter<VBuffer<double>> CreateGetter(DataViewRowCursor input, string inputCol)
+            {
+                ValueGetter<double> srcGetter = input.GetGetter<double>(input.Schema[inputCol]);
+                ValueGetter<VBuffer<double>> getter =
+                    (ref VBuffer<double> dst) =>
+                    {
+                        double src = default;
+                        srcGetter(ref src);
+                        var result = VBufferEditor.Create(ref dst, _outputLength);
+                        for (int i = 0; i < _outputLength; ++i)
+                        {
+                            result.Values[i] = _results[input.Position % _batchSize][i];
+                        }
+                        dst = result.Commit();
+                    };
+                return getter;
+            }
+        }
+
+        public sealed class SrCnnEntireModeler
```

Comment:
>public [](start = 8, length = 6)

Can this be private?

Reply:
I agree. Ideally this should be private

---
In reply to: [426413214](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426413214) [](ancestors = 426413214)

Reply:
This class is similar as Batch class, used by protect methods as parameter in SrCnnBatchAnomalyDetector, if make them private, the accessasiblity will not be aligned and there will be compile errors. I've made them internal.

---
In reply to: [426828580](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426828580) [](ancestors = 426828580,426413214)